### PR TITLE
Merge cosmos v4 to master

### PR DESCRIPTION
**Pull request by kushagraThapar**
* Merges v4 to master  - WITH RETAINING HISTORY
* Java code v4 to local branch merge is successful.
* Updated log4j version code change applied. (This was updated in v3, so had to pull this in the merge)
* Guava dependency reduced to 25-0.jre. (Didn't want to do this, but had to because of merge)
* Client - Track - 2 most of the changes are done. (adding cosmos sdk to various parent poms pom.client.xml, mandatory to have unit tests running on java 8 and java 11 - for all three windows, mac, linux)
* Dependency section updated for cosmos sdk. (eng/versioning/versions update).
* Skipped revapi 
* Skipped README
* Skipped Changelog
* Parent pom removal
* jacoco test coverage added (currently at 10%)

### CODE REVIEW

**Comment by moderakh** [comment]
azure-core brings in boringssl, rntbd has a known bug wigh tc-native and boringssl, please exclude boringssl till this bug is fixed.

File: sdk/cosmos/README.md [hunk 1]
```diff
@@ -1,296 +0,0 @@
-﻿# Azure CosmosDB Client Library for Java
```
Comment (weshaggard) [inline sdk/cosmos/README.md]: FWIW we still recommend having a root level README file that describes the set of packages in the folder. Something like https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/storage/README.md

Comment (kushagraThapar) [reply sdk/cosmos/README.md]: Sure, we can add it. That would be helpful to the users.

Comment (kushagraThapar) [reply sdk/cosmos/README.md]: Probably not as part of this PR, once we improve the content of other READMEs, may be then.

**Comment by moderakh** [comment]
I think log4j2 config is invalid:

see this:
```bash
2020-02-21T00:22:47.9239593Z 2020-02-21 00:22:47,915 TestNG-method=retries-1 ERROR An exception occurred processing Appender STDOUT java.lang.NullPointerException
2020-02-21T00:22:47.9261276Z 	at org.apache.logging.log4j.core.impl.ThrowableProxy.toExtendedStackTrace(ThrowableProxy.java:711)
2020-02-21T00:22:47.9262826Z 	at org.apache.logging.log4j.core.impl.ThrowableProxy.<init>(ThrowableProxy.java:138)
2020-02-21T00:22:47.9264232Z 	at org.apache.logging.log4j.core.impl.ThrowableProxy.<init>(ThrowableProxy.java:122)
2020-02-21T00:22:47.9265065Z 	at org.apache.logging.log4j.core.impl.MutableLogEvent.getThrownProxy(MutableLogEvent.java:351)
2020-02-21T00:22:47.9266392Z 	at org.apache.logging.log4j.core.pattern.ExtendedThrowablePatternConverter.format(ExtendedThrowablePatternConverter.java:64)
2020-02-21T00:22:47.9267946Z 	at org.apache.logging.log4j.core.pattern.PatternFormatter.format(PatternFormatter.java:38)
2020-02-21T00:22:47.9269010Z 	at org.apache.logging.log4j.core.layout.PatternLayout$PatternSerializer.toSerializable(PatternLayout.java:334)
2020-02-21T00:22:47.9270154Z 	at org.apache.logging.log4j.core.layout.PatternLayout.toText(PatternLayout.java:233)
2020-02-21T00:22:47.9271010Z 	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:218)
2020-02-21T00:22:47.9272115Z 	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:58)
2020-02-21T00:22:47.9279325Z 	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.directEncodeEvent(AbstractOutputStreamAppender.java:177)
2020-02-21T00:22:47.9285981Z 	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:170)
2020-02-21T00:22:47.9320315Z 	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:161)
2020-02-21T00:22:47.9321635Z 	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
2020-02-21T00:22:47.9322539Z 	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
2020-02-21T00:22:47.9340621Z 	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
2020-02-21T00:22:47.9341688Z 	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
2020-02-21T00:22:47.9346898Z 	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:464)
2020-02-21T00:22:47.9373348Z 	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:448)
2020-02-21T00:22:47.9374481Z 	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:431)
2020-02-21T00:22:47.9375531Z 	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:406)
2020-02-21T00:22:47.9419886Z 	at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
2020-02-21T00:22:47.9420996Z 	at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:146)
2020-02-21T00:22:47.9421858Z 	at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2170)
2020-02-21T00:22:47.9453107Z 	at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2125)
2020-02-21T00:22:47.9454242Z 	at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2108)
2020-02-21T00:22:47.9455112Z 	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2013)
2020-02-21T00:22:47.9517596Z 	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1882)
2020-02-21T00:22:47.9519325Z 	at org.apache.logging.slf4j.Log4jLogger.warn(Log4jLogger.java:254)
2020-02-21T00:22:47.9521207Z 	at com.azure.cosmos.implementation.ResourceThrottleRetryPolicy.shouldRetry(ResourceThrottleRetryPolicy.java:56)
```

from the test CI:
https://dev.azure.com/azure-sdk/29ec6040-b234-4e31-b139-33dc4287b756/_apis/build/builds/266358/logs/113

**Comment by kushagraThapar** [comment]
> I think log4j2 config is invalid:
> 
> see this:
> 
> ```shell
> 2020-02-21T00:22:47.9239593Z 2020-02-21 00:22:47,915 TestNG-method=retries-1 ERROR An exception occurred processing Appender STDOUT java.lang.NullPointerException
> 2020-02-21T00:22:47.9261276Z 	at org.apache.logging.log4j.core.impl.ThrowableProxy.toExtendedStackTrace(ThrowableProxy.java:711)
> 2020-02-21T00:22:47.9262826Z 	at org.apache.logging.log4j.core.impl.ThrowableProxy.<init>(ThrowableProxy.java:138)
> 2020-02-21T00:22:47.9264232Z 	at org.apache.logging.log4j.core.impl.ThrowableProxy.<init>(ThrowableProxy.java:122)
> 2020-02-21T00:22:47.9265065Z 	at org.apache.logging.log4j.core.impl.MutableLogEvent.getThrownProxy(MutableLogEvent.java:351)
> 2020-02-21T00:22:47.9266392Z 	at org.apache.logging.log4j.core.pattern.ExtendedThrowablePatternConverter.format(ExtendedThrowablePatternConverter.java:64)
> 2020-02-21T00:22:47.9267946Z 	at org.apache.logging.log4j.core.pattern.PatternFormatter.format(PatternFormatter.java:38)
> 2020-02-21T00:22:47.9269010Z 	at org.apache.logging.log4j.core.layout.PatternLayout$PatternSerializer.toSerializable(PatternLayout.java:334)
> 2020-02-21T00:22:47.9270154Z 	at org.apache.logging.log4j.core.layout.PatternLayout.toText(PatternLayout.java:233)
> 2020-02-21T00:22:47.9271010Z 	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:218)
> 2020-02-21T00:22:47.9272115Z 	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:58)
> 2020-02-21T00:22:47.9279325Z 	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.directEncodeEvent(AbstractOutputStreamAppender.java:177)
> 2020-02-21T00:22:47.9285981Z 	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:170)
> 2020-02-21T00:22:47.9320315Z 	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:161)
> 2020-02-21T00:22:47.9321635Z 	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
> 2020-02-21T00:22:47.9322539Z 	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
> 2020-02-21T00:22:47.9340621Z 	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
> 2020-02-21T00:22:47.9341688Z 	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
> 2020-02-21T00:22:47.9346898Z 	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:464)
> 2020-02-21T00:22:47.9373348Z 	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:448)
> 2020-02-21T00:22:47.9374481Z 	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:431)
> 2020-02-21T00:22:47.9375531Z 	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:406)
> 2020-02-21T00:22:47.9419886Z 	at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
> 2020-02-21T00:22:47.9420996Z 	at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:146)
> 2020-02-21T00:22:47.9421858Z 	at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2170)
> 2020-02-21T00:22:47.9453107Z 	at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2125)
> 2020-02-21T00:22:47.9454242Z 	at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2108)
> 2020-02-21T00:22:47.9455112Z 	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2013)
> 2020-02-21T00:22:47.9517596Z 	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1882)
> 2020-02-21T00:22:47.9519325Z 	at org.apache.logging.slf4j.Log4jLogger.warn(Log4jLogger.java:254)
> 2020-02-21T00:22:47.9521207Z 	at com.azure.cosmos.implementation.ResourceThrottleRetryPolicy.shouldRetry(ResourceThrottleRetryPolicy.java:56)
> ```
> 
> from the test CI:
> https://dev.azure.com/azure-sdk/29ec6040-b234-4e31-b139-33dc4287b756/_apis/build/builds/266358/logs/113

Yes, I saw that error too. 

@conniey    - Do you know what could be causing this ?

**Comment by conniey** [comment]
@kushagraThapar Does this test run a mock?
https://stackoverflow.com/a/47509211

File: eng/pipelines/templates/stages/cosmos-sdk-client.yml [hunk 2]
```diff
@@ -5,21 +5,54 @@ parameters:
...
     jobs:
-    - template: ../jobs/archetype-sdk-data.yml
+    - template: ../jobs/archetype-sdk-client.yml
       parameters:
         ServiceDirectory: ${{parameters.ServiceDirectory}}
         SDKType: ${{parameters.SDKType}}
         Artifacts: ${{parameters.Artifacts}}
         TestMatrix:
-          Windows - java8:
+          Linux - Java 8:
```
Comment (JimSuplizio) [inline eng/pipelines/templates/stages/cosmos-sdk-client.yml]: It's worth mentioning here that this test matrix, for the unit tests, brings Cosmos v4 inline with the test matrix used for client track (Linux/Mac/Windows with Java versions 8 and 11). The "from source" build is a special run that was added a while back that manipulates the versions such that all track client track dependencies are built from the latest source. This is a safeguard to ensure compatibility with changes being made to inter track dependencies.

File: eng/versioning/external_dependencies.txt [hunk 3]
```diff
@@ -171,11 +171,20 @@ applicationinsights_com.microsoft.azure:azure-client-runtime;1.6.3
 applicationinsights_com.microsoft.azure:azure-client-authentication;1.6.3
 
 # everything under sdk\cosmos
-cosmos_org.mockito:mockito-core;1.10.19
+cosmos_com.fasterxml.jackson.module:jackson-module-afterburner;2.10.1
 cosmos_com.google.guava:guava;25.0-jre
 cosmos_io.dropwizard.metrics:metrics-core;4.1.0
-cosmos_io.dropwizard.metrics:metrics-jvm;4.1.0
 cosmos_io.dropwizard.metrics:metrics-graphite;4.1.0
+cosmos_io.dropwizard.metrics:metrics-jvm;4.1.0
```
Comment (JimSuplizio) [inline eng/versioning/external_dependencies.txt]: /CC @joshfree @alzimmermsft 
Alan and I have chatted about these netty dependencies. Right now the Cosmos io.netty dependencies are ahead of the common versions. We should look and see if we can bring the common dependencies up to these versions.

File: pom.client.xml [hunk 4]
```diff
@@ -428,6 +428,10 @@
               *.implementation*:
               com.azure.tools.checkstyle*
             </excludePackageNames>
+            <!-- CosmosSkip - This is temporary, BridgeInternal gets converted to implementation -->
```
Comment (JimSuplizio) [inline pom.client.xml]: Right now, the goal is to get Comos v4 into master and successfully building in the pipelines which will unblock the development. There are still source changes that need to get made per the arch board review that will not be part of this PR. @kushagraThapar and I agreed that tagging things in the pom files, like plugin overrides or things that are temporary, with "CosmosSkip" so they can be easily identified. It is my understanding that everything tagged with CosmosSkip is going to be attached to a work item which is to be completed before GA.

File: eng/versioning/external_dependencies.txt (hunk 3 above)
Comment (alzimmermsft) [reply eng/versioning/external_dependencies.txt]: I'm good with upgrading to these newer versions if we get a few good test runs in with them.

**Pull request merged**